BUG: Work around past and present PEP3118 issues in ctypes

eric-wieser opened the pull request:
Fixes gh-10528
Fixes gh-10978
Fixes gh-11150

It seems we already have some `ctype -> dtype` code that we can fall back on in the frequent cases where ctypes give the wrong pep3118 buffer.

---


This is kind of unfortunate - I was hoping to implement `ctype -> dtype` in terms of the PEP3118 buffer code (https://github.com/numpy/numpy/issues/10533#issuecomment-363681273), but it seems that we have to fall back on the reverse here

eric-wieser commented:
Test results look pretty damning

eric-wieser committed:
MAINT: Don't steal a reference

eric-wieser committed:
MAINT: Extract helper function

charris commented on numpy/core/_internal.py:
```diff
@@ -756,3 +756,12 @@ def _ufunc_doc_signature_formatter(ufunc):
+    # determine if an object comes from ctypes, in order to work around
+    # a bug in the buffer protocol for those objects, bpo-10746
+    try:
+        return 'ctypes' in type(obj).__mro__[-2].__module__
```
`__mro__` is only available for new style classes, might be a problem in Python 2.7.

charris replied:
Although I suspect anything that inherits from ctypes is going to be a new style class, so this should be OK. Maybe a note?

eric-wieser replied:
Yep, it seemed that ctypes classes are always new-style

charris commented on numpy/core/_internal.py:
```diff
@@ -756,3 +756,12 @@ def _ufunc_doc_signature_formatter(ufunc):
+    try:
+        return 'ctypes' in type(obj).__mro__[-2].__module__
+    except Exception:
+        return False
```
File needs newline at end.

charris commented on numpy/core/src/multiarray/ctors.c:
```diff
@@ -1302,7 +1303,47 @@ PyArray_New(PyTypeObject *subtype, int nd, npy_intp *dims, int type_num,
+        }
+    }
+    else {
+        descr = PyArray_DescrNewFromType(NPY_STRING);
```
Maybe note on string assumption.

charris commented on numpy/core/src/multiarray/ctors.c:
```diff
@@ -1315,27 +1356,73 @@ _array_from_buffer_3118(PyObject *memoryview)
     npy_intp shape[NPY_MAXDIMS], strides[NPY_MAXDIMS];
 
     view = PyMemoryView_GET_BUFFER(memoryview);
```
Is is safe to assume that `memoryview` is actually a memoryview?

eric-wieser replied:
Yes, because this function is only ever called with one, I think

charris commented on numpy/core/src/multiarray/ctors.c:
```diff
@@ -1315,27 +1356,73 @@ _array_from_buffer_3118(PyObject *memoryview)
+        if (descr == NULL) {
+            return NULL;
+        }
+        if (descr->elsize != view->len) {
```
I'm confused here as `view->len` should be the total size of the memory. What am I missing?

eric-wieser replied:
We're creating an 0d array below with one (compound) element - so the dtype should span the total size of the memory

charris replied:
A comment somewhere would help, as that is a bit unexpected. At least, I didn't expect it :)

charris commented:
Couple of nits and a question.

eric-wieser committed:
BUG: Work around bugs in pep3118 ctypes by falling back on np.dtype(type(ctype_obj))

Fixes gh-10528
Fixes gh-10978
Fixes gh-11150

A warning is thrown when ctypes misbehaves, in order to not hide issues that need fixing upstream

eric-wieser commented:
Comments improved
